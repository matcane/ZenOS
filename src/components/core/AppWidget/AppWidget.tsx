import { Pressable, PressableProps, Image } from "react-native";

import { APPS_ICON, TApp } from "@/constants/core";
import { baseStyle } from "@/styles/baseStyle";
import { coreStyles } from "@/styles/core";
import { Colors } from "@/theme";

import ThemedText from "../ThemedText/ThemedText";

const { textCenter } = baseStyle;
const { container } = coreStyles;

type AppWidgetProps = PressableProps & {
  app: TApp;
  itemWidth: number;
  onlyIcon?: boolean;
};

export default function AppWidget({ app, itemWidth, onlyIcon, ...rest }: AppWidgetProps) {
  const additionalStyles = { color: Colors.dark.text, width: itemWidth };
  return (
    <Pressable testID="pressable" style={[container, { width: itemWidth }]} {...rest}>
      <Image source={APPS_ICON[app.icon]} testID="icon" />
      {onlyIcon ? undefined : (
        <ThemedText style={[textCenter, additionalStyles]} numberOfLines={1}>
          {app.name}
        </ThemedText>
      )}
    </Pressable>
  );
}
