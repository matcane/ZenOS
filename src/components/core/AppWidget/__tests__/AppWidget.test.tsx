import { render, fireEvent } from "@testing-library/react-native";

import AppWidget from "../AppWidget";

const app = {
  icon: "@assets/icons/icons8-clock.webp",
  name: "App Name",
  slug: "app-name",
  version: "0.0.1",
};

const ITEM_WIDTH = 100;

describe("AppWidget component", () => {
  it("renders correctly", () => {
    const tree = render(<AppWidget app={app} itemWidth={ITEM_WIDTH} />);

    expect(tree).toMatchSnapshot();
  });

  it("renders the icon", () => {
    const { getByTestId } = render(<AppWidget app={app} itemWidth={ITEM_WIDTH} />);
    const icon = getByTestId("icon");
    expect(icon).toBeTruthy();
  });

  it("renders the app name when onlyIcon is false", () => {
    const { getByText } = render(<AppWidget app={app} itemWidth={ITEM_WIDTH} />);
    const appName = getByText(app.name);
    expect(appName).toBeTruthy();
  });

  it("does not render the app name when onlyIcon is true", () => {
    const { queryByText } = render(<AppWidget app={app} itemWidth={ITEM_WIDTH} onlyIcon />);
    const appName = queryByText(app.name);
    expect(appName).toBeNull();
  });

  it("calls the onPress event when pressed", () => {
    const onPress = jest.fn();
    const { getByTestId } = render(
      <AppWidget app={app} itemWidth={ITEM_WIDTH} onPress={onPress} />,
    );
    const pressable = getByTestId("pressable");
    fireEvent.press(pressable);
    expect(onPress).toHaveBeenCalledTimes(1);
  });

  it("renders with the correct styles", () => {
    const { getByTestId } = render(<AppWidget app={app} itemWidth={ITEM_WIDTH} />);
    const container = getByTestId("pressable");
    const style = container.props.style;
    expect(style).toEqual(expect.arrayContaining([expect.objectContaining({ width: ITEM_WIDTH })]));
  });
});
