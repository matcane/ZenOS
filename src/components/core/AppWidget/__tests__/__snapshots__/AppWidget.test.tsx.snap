// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppWidget component renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "padding": 20,
      },
      {
        "width": 100,
      },
    ]
  }
  testID="pressable"
>
  <Image
    source={1}
    testID="icon"
  />
  <Text
    numberOfLines={1}
    style={
      [
        {
          "color": "#212121",
        },
        {
          "fontSize": 16,
        },
        [
          {
            "textAlign": "center",
          },
          {
            "color": "#ffffff",
            "width": 100,
          },
        ],
      ]
    }
  >
    App Name
  </Text>
</View>
`;
